### ISSUE TITLE
Support for collection-doi in APA and BibTeX formatter

### ISSUE THREAD

**Issue by lnielsen** [citation-file-format/ruby-cff#41]
*Just a suggestion for improvement*

If you add the ``collection-doi`` field it's not used by the APA and BibTex formatters. It would be great if there was some resolution mechanism implemented that would

- Use 'doi' if it exists.
- Use 'collection-doi' if 'doi' not present.
- Use 'url' or some other identifier if neither of the above exist.

This is to ensure that at least one identifier makes it into a citation string as this greatly helps with later citation tracking. 

cc @arfon

**Comment by hainesr** [comment]
Hi @lnielsen, many thanks for this note; this is really useful insight!

I will look at using `collection-doi` if we don't have a `doi` as you suggest.

Looking at the current code it looks like we insert a URI if it's there - even if we also have a DOI. And for the URI we prefer the `repository_code` over `url` if provided. Does this seem reasonable?

And if there is no URI at all, is there anything else we could usefully provide as an identifier do you think?

Thanks.

**Comment by arfon** [comment]
> And if there is no URI at all, is there anything else we could usefully provide as an identifier do you think?

We presumably want to use a Software Heritage identifier too if there is one? e.g., https://github.com/citation-file-format/citation-file-format/blob/main/README.md#exemplary-uses-3

**Comment by lnielsen** [comment]
> Looking at the current code it looks like we insert a URI if it's there - even if we also have a DOI. 

I've tried to check what the Citation Style Language does for the APA style. As far as I can decode it only includes the DOI and not both DOI/URL. See the [APA style file](https://github.com/citation-style-language/styles/blob/2cd6364de42525f99d00b99d790603643ec67a52/apa.csl#L1549-L1569)

From a citation tracking point of view one identifier is preferable (i.e. no ambiguity). From a user point of view, I'm not sure - I think there's a certain value in having the URL there as well.

> And for the URI we prefer the ``repository_code`` over ``url`` if provided. Does this seem reasonable?

Yes, fully agree.

> We presumably want to use a Software Heritage identifier too if there is one? e.g., https://github.com/citation-file-format/citation-file-format/blob/main/README.md#exemplary-uses-3

Yep agree, an SWH identifier can be used as well - probably via a resolver link:

```
# https://docs.softwareheritage.org/devel/swh-model/persistent-identifiers.html#software-heritage-resolver
https://archive.softwareheritage.org/<identifier>
```

**Comment by hainesr** [comment]
I just realised last night (#43) that the ruby library doesn't have the `identifiers` type yet - it was added to the CFF spec recently. I'll add that this week. Then where would we use a resolver link for a SWH identifier in APA/BibTeX?

Also, I was looking at the `collection-doi` field, and it isn't present at the top level for the software object itself. It's only potentially present in a reference. So I don't know if it makes sense to pull that to the top level, unless I've missed something.

**Comment by lnielsen** [comment]
> Also, I was looking at the collection-doi field, and it isn't present at the top level for the software object itself. It's only potentially present in a reference. So I don't know if it makes sense to pull that to the top level unless I've missed something.

You're completely right - I completely missed that it was related to a reference and not top-level key. So, yes it doesn't make sense to use it.

**Comment by hainesr** [comment]
Thanks @lnielsen, @arfon.

So I think then the workflow for choosing a `doi` would be:

1. Use a top-level `doi` if provided, else
2. Use a `doi` from the top-level `identities` field (subject to #43)

And for choosing a URI:

1. Use `repository_code` if present, else
2. Use `url` if present, else
3. Use a resolvable SWH URI from the top-level `identities` field if present

Or would we swap 2 and 3 for the URI? Or would we bump SWH to 1?

**Comment by lnielsen** [comment]
I would go with moving the SWH to the ``doi`` just because it's a more stable identifier than the URL and users have to manually add it to the CFF file  thus making an active choice to promote the SWH identifier.

**Comment by hainesr** [comment]
Is it acceptable to use an SWH ID as a DOI? As in, can anything that is considered archival be used as a DOI, or should they be of the `10.5281/zenodo.1003150` form?

**Comment by hainesr** [comment]
Ah, an SWH ID won't pass validation as a DOI.

**Comment by mfenner** [comment]
SWH ID is not a DOI, but another kind of persistent identifier. It would be desirable to have both if possible as they address different use cases. Most DOIs (about 85%) for software are from Zenodo, but it is not a requirement.
